### PR TITLE
Add Counter and Histogram Aggregators 

### PR DESCRIPTION
**Pull request by ankit-bhargava**
This PR builds out SDK support for the [Counter and Histogram Aggregators](https://github.com/jmacd/opentelemetry-specification/blob/jmacd/draft_metric_sdk_spec/specification/metrics/sdk.md#aggregator-implementations) which combine updates to metric instruments in meaningful values.  Each aggregator stores exportable values in a vector format for ease of use further down the metrics data pipeline.   All aggregators are derived from a base Aggregator abstract class which holds functions and data relevant to all aggregators.  Additional data structures, such as a vector for histogram boundaries, are added on a case by case basis.  Aggregators are dependent on the Metric Instruments API  (pending approval in [PR # 161](https://github.com/open-telemetry/opentelemetry-cpp/pull/161)) and **will not compile without those files.**

Also included are tests for each aggregator which validate the correctness of aggregation in both single-threaded and concurrent situations.

Note: We templated aggregators to support a wider array of scalar types.  This templating prevents us from using out of line declarations as we normally would.  As a result, we left all implementation code in the header file.  If there are any suggestions to remedy this we will gladly make the change.  

This PR is part of a series of PRs to add a full implementation of metrics API and SDK support.

### CODE REVIEW

File: sdk/include/opentelemetry/sdk/metrics/aggregator/aggregator.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+#include "opentelemetry/version.h"
+#include "opentelemetry/metrics/instrument.h"
+
+#include <variant>
+#include <vector>
+#include <mutex>
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/aggregator/aggregator.h:8]: nit - is there a reason these are not alphabetically sorted? If not, I would suggest to sort them.

File: sdk/include/opentelemetry/sdk/metrics/aggregator/aggregator.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+ */
+template<typename T>
+class Aggregator
+{
+public:
+    
+    Aggregator() = default;
+    
+    /**
+     * Recieves a captured value from the instrument and applies it to the current aggregator value.
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/aggregator/aggregator.h:32]:
```suggestion
     * Receives a captured value from the instrument and applies it to the current aggregator value.
```

File: sdk/include/opentelemetry/sdk/metrics/aggregator/counter_aggregator.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    /**
+     * Recieves a captured value from the instrument and applies it to the current aggregator value.
+     *
+     * @param val, the raw value used in aggregation
+     * @return none
+     */
+    void update(T val) override
+    {
+        this->mu_.lock();
+        this->values_[0] += val;  // atomic operation
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/aggregator/counter_aggregator.h:41]: For int and long, is it possible to spin + CAS (compare and swap) to achieve better perf?
Not blocking this PR though.

File: sdk/include/opentelemetry/sdk/metrics/aggregator/counter_aggregator.h [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+     *
+     * @param other, the aggregator with merge with
+     * @return none
+     */
+    void merge(CounterAggregator other)
+    {
+        if (this->kind_ == other.kind_)
+        {
+            this->mu_.lock();
+            this->values_[0] += other.values_[0];  // atomic operation afaik
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/aggregator/counter_aggregator.h:70]: curious, what does `// atomic operation afaik` mean?

File: sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+     * Recieves a captured value from the instrument and inserts it into the current histogram counts.
+     * Depending on the use case, linear search or binary search based implementations may be preferred.
+     * In a uniformly distributed dataset, linear search outperforms binary search until 512 buckets.
+     *
+     * @param val, the raw value used in aggregation
+     * @return none
+     */
+    void update(T val) override
+    {
+
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:51]: nit: remove empty line?

File: sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+     * Constructor for the histogram aggregator.  A sorted vector of boundaries is expected and boundaries are
+     * doubles regardless of the aggregator's templated data type.
+     *
+     * Sum is stored in values_[0]
+     * Count is stored in position_[1]
+     */
+    HistogramAggregator(metrics_api::BoundInstrumentKind kind, std::vector<double> boundaries)
+    {
+        this->kind_ = kind;
+        boundaries_   = boundaries;
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:35]: Do we want to validate if the boundaries is monotonic?

File: sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        for (int i = 0; i < boundaries_.size(); i++)
+        {
+            if (val < boundaries_[i]) // concurrent read is thread-safe
+            {
+                bucketID = i;
+                break;
+            }
+        }
+        
+        // Alternate implementation with binary search
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:62]: It'll be helpful to put a comment on the design decision - e.g. why would we choose linear search vs. binary search.

File: sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+     */
+    void merge(HistogramAggregator other)
+    {
+        this->mu_.lock();
+        this->values_[0] += other.values_[0];
+        this->values_[1] += other.values_[1];
+        
+        for (int i = 0; i < bucketCounts_.size(); i++)
+        {
+            bucketCounts_[i] += other.bucketCounts_[i];
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:104]: Who should be responsible of making sure the `other.bucketCounts` won't overflow? Do we in general want to check the buckets are the same before proceeding?

File: sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    void checkpoint() override
+    {
+        this->checkpoint_ = this->values_;
+        this->values_[0]=0;
+        this->values_[1]=0;
+    }
+    
+    /**
+     * Merges the values of two aggregators in a semantically accurate manner.
+     * A histogram aggregator can only be merged with another histogram aggregatos with the same boudnaries.
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:89]: Theoretically we could merge a more accurate histogram into a less accurate one (e.g. merge [0, 1, 10, 100, 1000) into [0, 10, 1000). Probably not something we should cover here.

File: sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h (hunk 8 above)
Comment (ankit-bhargava) [reply sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:104]: This is a good suggestion.  When you say " `other.bucketCounts` won't overflow," are you referring to a scenario in which adding the values together would exceed the capacity of the data type?

Comment (reyang) [reply sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:104]: For example, if `other.bucketCounts_.size() <= i`.

Comment (reyang) [reply sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:104]: There are two separate things:
1. index overflow while accessing the bucketCounts_.
2. value overflow while performing `+=`.

1 is more scary.

File: sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h (hunk 6 above)
Comment (ankit-bhargava) [reply sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:40]: Added a check which throws an invalid_argument exception if provided boundaries are not sorted

File: sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h (hunk 7 above)
Comment (ankit-bhargava) [reply sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:71]: Included some background on the topic above the update function -- let me know if this is not enough and I'll gladly add a more thorough explanation.

File: sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h (hunk 8 above)
Comment (ankit-bhargava) [reply sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:122]: I preempted the possibility of an index out of bounds issue by ensuring that the boundary vectors of both aggregators are identical.  

For the value overflow, I can think of some hacky ways to do it, for example if abs(a)+abs(b) < 0 to indicate a wraparound.  I could also subtract from a variable set to int_max.  I feel like there has to be a native c++ tool/function which does this but I can't seem to find it...

**Comment by ankit-bhargava** [comment]
Addressed comments and rebased.  This has a dependency on PR #161.

**Comment by ankit-bhargava** [comment]
Failing Bazel noexcept test although I was under the impression we were allowed to throw exceptions in SDK components?  Unsure why the exporter proto CMake test is failing.

**Comment by ankit-bhargava** [comment]
Passing all tests except for "exporter proto" which is an external issue.

File: sdk/include/opentelemetry/sdk/metrics/aggregator/aggregator.h [hunk 10]
```diff
@@ -0,0 +1,164 @@
+#pragma once
+
+#include <mutex>
+#include "opentelemetry/core/timestamp.h"
+#include "opentelemetry/metrics/instrument.h"
+#include "opentelemetry/version.h"
+#include <vector>
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/aggregator/aggregator.h:7]: Move this line above (right below <mutex>).

File: sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+    /**
+     * Constructor for the histogram aggregator.  A sorted vector of boundaries is expected and boundaries are
+     * doubles regardless of the aggregator's templated data type.
+     *
+     * Sum is stored in values_[0]
+     * Count is stored in position_[1]
+     */
+    HistogramAggregator(metrics_api::InstrumentKind kind, std::vector<double> boundaries)
+    {
+        if (!std::is_sorted(boundaries.begin(),boundaries.end())){
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:34]:
```suggestion
        if (!std::is_sorted(boundaries.begin(), boundaries.end())){
```

File: sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+        bucketCounts_ = std::vector<int>(boundaries_.size() + 1, 0);
+        bucketCounts_ckpt_ = std::vector<int>(boundaries_.size() + 1, 0);
+    }
+    
+    /**
+     * Recieves a captured value from the instrument and inserts it into the current histogram counts.
+     *
+     * Depending on the use case, a linear search or binary search based implementation may be preferred.
+     * In uniformly distributed datasets, linear search outperforms binary search until 512 buckets.  However,
+     * if the distribution is strongly skewed right ( for example server latency where most values may be <10ms
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h:55]:
```suggestion
     * if the distribution is strongly skewed right (for example server latency where most values may be <10ms
```

**Comment by reyang** [comment]
@ankit-bhargava would you fix the CI/Format issue?

**Comment by ankit-bhargava** [comment]
@reyang 
Fixed all the formatting issues and rebased.  Seeing two runtime_context files which should not be part of this PR and I'm not sure why they've been added and if they're necessary? Feel free to delete these and merge (or I can delete them).

File: sdk/include/opentelemetry/sdk/metrics/aggregator/aggregator.h (hunk 1 above)
Comment (ankit-bhargava) [reply sdk/include/opentelemetry/sdk/metrics/aggregator/aggregator.h:3]: the format tools seem to reorder these

**Pull request merged**